atlasdb-config/src/main/java/com/palantir/atlasdb/factory/InitialisingTransactionManager.java

```diff
@@ -0,0 +1,667 @@
+            boolean allowHiddenTableAccess,
+            String userAgent) {
+
+        atlasFactory = new ServiceDiscoveringAtlasSupplier(config);
+        this.config = config;
+        this.runtimeConfigSupplier = runtimeConfigSupplier;
+        this.schemas = schemas;
+        this.env = env;
+        this.lockServerOptions = lockServerOptions;
+        this.allowHiddenTableAccess = allowHiddenTableAccess;
+        this.userAgent = userAgent;
+
+        LockRequest.setDefaultLockTimeout(
+                SimpleTimeDuration.of(config.getDefaultLockTimeoutSeconds(), TimeUnit.SECONDS));
+
+        registerEndpoints();
+
+        try {
+            initialise();
+        } catch (Throwable th) {
+            log.info("Synchronous initialisation failed, initialisation will be done asynchronously", th);
+            initialiseAsync();
+        }
+    }
+
+    private void registerEndpoints() {
+        env.register(remoteLockService);
+        env.register(timestampService);
+        if (config.leader().isPresent()) {
+            localPaxosServices = Leaders.createAndRegisterLocalServices(env, config.leader().get(), userAgent);
+        }
+        env.register(persistentLockService);
+        env.register(new CheckAndSetExceptionMapper());
+        env.register(sweeperService);
+    }
+
+    private void initialiseAsync() {
+        runAsync.accept(() -> {
+            while (uninitialised()) {
+                try {
+                    initialise();
+                } catch (Throwable th) {
+                    log.info("Async initialisation failed, retrying in {} seconds.", RETRY_AFTER_SECONDS, th);
+                    Uninterruptibles.sleepUninterruptibly(RETRY_AFTER_SECONDS, TimeUnit.SECONDS);
+                }
+            }
+        });
+    }
+
+    private void initialise() {
+        KeyValueService rawKvs = atlasFactory.getKeyValueService();
+
+        LockAndTimestampServices lockAndTimestampServices = createLockAndTimestampServices(
+                JavaSuppliers.compose(getTimestampClientConfigFunction(), runtimeConfigSupplier),
+                () -> LockServiceImpl.create(lockServerOptions),
+                atlasFactory::getTimestampService,
+                atlasFactory.getTimestampStoreInvalidator());
+
+        KeyValueService kvs = NamespacedKeyValueServices.wrapWithStaticNamespaceMappingKvs(rawKvs);
+        kvs = ProfilingKeyValueService.create(kvs, config.getKvsSlowLogThresholdMillis());
+        kvs = SweepStatsKeyValueService.create(kvs,
+                new TimelockTimestampServiceAdapter(lockAndTimestampServices.timelock()));
+        kvs = TracingKeyValueService.create(kvs);
+        kvs = AtlasDbMetrics.instrument(KeyValueService.class, kvs,
+                MetricRegistry.name(KeyValueService.class, userAgent));
+        kvs = ValidatingQueryRewritingKeyValueService.create(kvs);
+
+        TransactionTables.createTables(kvs);
+
+        persistentLockService.initialise(createPersistentLockService(kvs));
```

Comment:
I haven't drawn the dependency graph, but it looks like a lot of the components depend on the KVS, which is the main thing that can go wrong - so I don't think this will gain that much.